dotnet/src/SemanticKernel.Core/Functions/OrderedIAIServiceSelector.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.SemanticKernel.AI;
+using Microsoft.SemanticKernel.AI.TextCompletion;
+using Microsoft.SemanticKernel.Diagnostics;
+using Microsoft.SemanticKernel.Services;
+
+namespace Microsoft.SemanticKernel.Functions;
+
+/// <summary>
+/// Implementation of <see cref="IAIServiceSelector"/> that selects the AI service based on the order of the model settings.
+/// Uses the service id to select the preferred service provider and then returns the service and associated model settings.
+/// </summary>
+internal class OrderedIAIServiceSelector : IAIServiceSelector
+{
+    /// <inheritdoc/>
+    public (T?, AIRequestSettings?) SelectAIService<T>(IAIServiceProvider serviceProvider, List<AIRequestSettings>? modelSettings) where T : IAIService
+    {
+        if (modelSettings == null || modelSettings.Count == 0)
+        {
+            var service = serviceProvider.GetService<T>(null);
+            if (service != null)
+            {
+                return (service, modelSettings?.FirstOrDefault<AIRequestSettings>());
+            }
+        }
+        else
+        {
+            AIRequestSettings? defaultRequestSettings = null;
+            foreach (var model in modelSettings)
+            {
+                if (model.ServiceId is not null)
+                {
+                    var service = serviceProvider.GetService<T>(model.ServiceId);
+                    if (service != null)
+                    {
+                        return (service, model);
+                    }
+                }
+                else
+                {
+                    // First request settings with no service id is the default
+                    defaultRequestSettings ??= model;
+                }
+            }
+
+            if (defaultRequestSettings is not null)
+            {
+                return (serviceProvider.GetService<T>(null), defaultRequestSettings);
+            }
+        }
+
+        var names = string.Join("|", modelSettings.Select(model => model.ServiceId).ToArray());
+        throw new SKException($"Service of type {typeof(ITextCompletion)} and name {names ?? "<NONE>"} not registered.");
```

Comment:
I believe you wanted to include type of the T into the exception message and not of ITextCompletion?